be/src/agent/publish_version.cpp

```diff
@@ -0,0 +1,161 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include "publish_version.h"
+
+#include <bvar/bvar.h>
+
+#include "gutil/strings/join.h"
+#include "storage/data_dir.h"
+#include "storage/storage_engine.h"
+#include "storage/tablet.h"
+#include "util/monotime.h"
+#include "util/threadpool.h"
+
+namespace starrocks {
+
+const uint32_t PUBLISH_VERSION_SUBMIT_MAX_RETRY = 10;
+bvar::LatencyRecorder g_publish_latency("be", "publish");
+
+void flush_dirs(std::unordered_set<DataDir*>& affected_dirs) {
+    for (auto dir : affected_dirs) {
+        auto st = dir->get_meta()->flush();
+        if (!st.ok()) {
+            LOG(WARNING) << "failed to flush tablet meta, dir:" << dir->path() << " res:" << st;
+        }
+    }
+}
+
+struct TabletPublishVersionTask {
+    int64_t txn_id{0};
+    int64_t partition_id{0};
+    int64_t tablet_id{0};
+    int64_t version{0};
+    RowsetSharedPtr rowset;
+    Status st;
+    int64_t max_continuous_version{0}; // max continuous version after publish is done
+};
+
+void run_publish_version_task(ThreadPool& threadpool, const TAgentTaskRequest& publish_version_task,
+                              TFinishTaskRequest& finish_task, std::unordered_set<DataDir*>& affected_dirs) {
+    int64_t start_ts = MonotonicMillis();
+    auto publish_version_req = publish_version_task.publish_version_req;
+    int64_t transaction_id = publish_version_req.transaction_id;
+    size_t num_partition = publish_version_req.partition_version_infos.size();
+    size_t num_tablet = 0;
+    std::vector<std::map<TabletInfo, RowsetSharedPtr>> partitions(num_partition);
+    for (size_t i = 0; i < publish_version_req.partition_version_infos.size(); i++) {
+        StorageEngine::instance()->txn_manager()->get_txn_related_tablets(
+                transaction_id, publish_version_req.partition_version_infos[i].partition_id, &partitions[i]);
+        num_tablet += partitions[i].size();
+    }
+    std::vector<TabletPublishVersionTask> tablet_tasks(num_tablet);
+    size_t tablet_idx = 0;
+    for (size_t i = 0; i < publish_version_req.partition_version_infos.size(); i++) {
+        for (auto& itr : partitions[i]) {
+            auto& task = tablet_tasks[tablet_idx++];
+            task.txn_id = transaction_id;
+            task.partition_id = publish_version_req.partition_version_infos[i].partition_id;
+            task.tablet_id = itr.first.tablet_id;
+            task.version = publish_version_req.partition_version_infos[i].version;
+            task.rowset = std::move(itr.second);
+        }
+    }
+    std::mutex affected_dirs_lock;
+    for (size_t i = 0; i < tablet_tasks.size(); i++) {
+        uint32_t retry_time = 0;
+        Status st;
+        while (retry_time++ < PUBLISH_VERSION_SUBMIT_MAX_RETRY) {
+            st = threadpool.submit_func([&, i]() {
+                auto task = tablet_tasks[i];
```

Comment:
I think you want to change the task in tablet_tasks[i], so I think this should be a reference?

Reply:
right

Reply:
that's a serious bug